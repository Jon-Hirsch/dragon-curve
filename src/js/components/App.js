import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      iterations: '12',
      iterationToTrack: '0',
      colorMode: 'rainbow',
      numberOfCurves: 1,
      useAnimation: true
    };

    this.drawCurve = this.drawCurve.bind(this);
  }

  drawCurve() {
    const {
      iterations,
      iterationToTrack,
      colorMode,
      numberOfCurves,
      useAnimation
    } = this.state;

    this.props.draw(
      isNaN(iterations) ? 12 : +iterations,
      isNaN(iterationToTrack) ? 0 : +iterationToTrack,
      colorMode,
      numberOfCurves,
      useAnimation
    );
  }

  updateField(field, value) {
    this.setState(state => ({
      [field]: value
    }));
  }

  render() {
    return (
      <div className="dragon-curve-controls">
        <div className="button-container">
          <button onClick={this.drawCurve}>Draw Curve</button>
        </div>
        <div className="dragon-curve-controls-row">
          <div>
            <label>
              Iterations:{' '}
              <input
                type="text"
                value={this.state.iterations}
                className="dragon-curve-iterations"
                onChange={event =>
                  this.updateField('iterations', event.target.value)}
              />
            </label>
          </div>
          <div>
            Color Mode:{' '}
            <select
              value={this.state.colorMode}
              className="dragon-curve-color-mode"
              onChange={event =>
                this.updateField('colorMode', event.target.value)}
            >
              <option value="rainbow">Rainbow</option>
              <option value="contrast">Contrast</option>
              <option value="black">Black</option>
            </select>
          </div>
          <div>
            <label>
              Use Animation:{' '}
              <input
                type="checkbox"
                checked={this.state.useAnimation}
                className="dragon-curve-animation-checkbox"
                onChange={event =>
                  this.updateField('useAnimation', event.target.checked)}
              />
            </label>
          </div>
        </div>
        <div className="dragon-curve-controls-row">
          <div>
            Number of Curves:{' '}
            <select
              value={this.state.numberOfCurves}
              className="dragon-curve-number-of-curves"
              onChange={event =>
                this.updateField('numberOfCurves', event.target.value)}
            >
              <option value="1">1</option>
              <option value="2">2</option>
              <option value="3">3</option>
              <option value="4">4</option>
            </select>
          </div>
          <div>
            <label>
              Iteration to Track{' '}
              <input
                type="text"
                value={this.state.iterationToTrack}
                className="dragon-curve-iteration-to-track"
                onChange={event =>
                  this.updateField('iterationToTrack', event.target.value)}
              />
            </label>
          </div>
          <div />
        </div>
      </div>
    );
  }
}

App.propTypes = {
  draw: PropTypes.func.isRequired
};
