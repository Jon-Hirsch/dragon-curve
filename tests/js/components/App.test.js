import React from 'react';
import App from '../../../src/js/components/App';

describe('App', () => {
  let shallowWrapper, draw;

  beforeEach(() => {
    draw = jest.fn();
    shallowWrapper = render();
  });

  function render() {
    return shallow(<App draw={draw} />);
  }

  test('should match snapshot', () => {
    expect(render().getElement()).toMatchSnapshot();
  });

  test('click the Draw Curve button calls draw', () => {
    shallowWrapper.find('button').simulate('click');
    expect(draw).toHaveBeenCalledWith(12, 0, 'rainbow', 1, true);
  });

  test('sends default values to draw if the values are not valid numbers', () => {
    shallowWrapper.setState(state => ({
      iterations: 'a',
      iterationToTrack: 'b'
    }));
    shallowWrapper.find('button').simulate('click');
    expect(draw).toHaveBeenCalledWith(12, 0, 'rainbow', 1, true);
  });

  test('updates state when the iterations input changes', () => {
    shallowWrapper
      .find('.dragon-curve-iterations')
      .simulate('change', { target: { value: '10' } });
    expect(shallowWrapper.state().iterations).toEqual('10');
  });

  test('updates state whenthe  colorMode dropdown changes', () => {
    shallowWrapper
      .find('.dragon-curve-color-mode')
      .simulate('change', { target: { value: 'black' } });
    expect(shallowWrapper.state().colorMode).toEqual('black');
  });

  test('updates state when the use animation checkbox changes', () => {
    shallowWrapper
      .find('.dragon-curve-animation-checkbox')
      .simulate('change', { target: { checked: false } });
    expect(shallowWrapper.state().useAnimation).toEqual(false);
  });

  test('updates state when the number of curves input changes', () => {
    shallowWrapper
      .find('.dragon-curve-number-of-curves')
      .simulate('change', { target: { value: '2' } });
    expect(shallowWrapper.state().numberOfCurves).toEqual('2');
  });

  test('updates state when the iteration to track input changes', () => {
    shallowWrapper
      .find('.dragon-curve-iteration-to-track')
      .simulate('change', { target: { value: '2' } });
    expect(shallowWrapper.state().iterationToTrack).toEqual('2');
  });
});
