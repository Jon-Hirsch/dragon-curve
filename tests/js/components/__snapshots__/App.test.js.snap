// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  className="dragon-curve-controls"
>
  <div
    className="button-container"
  >
    <button
      onClick={[Function]}
    >
      Draw Curve
    </button>
  </div>
  <div
    className="dragon-curve-controls-row"
  >
    <div>
      <label>
        Iterations:
         
        <input
          className="dragon-curve-iterations"
          onChange={[Function]}
          type="text"
          value="12"
        />
      </label>
    </div>
    <div>
      Color Mode:
       
      <select
        className="dragon-curve-color-mode"
        onChange={[Function]}
        value="rainbow"
      >
        <option
          value="rainbow"
        >
          Rainbow
        </option>
        <option
          value="contrast"
        >
          Contrast
        </option>
        <option
          value="black"
        >
          Black
        </option>
      </select>
    </div>
    <div>
      <label>
        Use Animation:
         
        <input
          checked={true}
          className="dragon-curve-animation-checkbox"
          onChange={[Function]}
          type="checkbox"
        />
      </label>
    </div>
  </div>
  <div
    className="dragon-curve-controls-row"
  >
    <div>
      Number of Curves:
       
      <select
        className="dragon-curve-number-of-curves"
        onChange={[Function]}
        value={1}
      >
        <option
          value="1"
        >
          1
        </option>
        <option
          value="2"
        >
          2
        </option>
        <option
          value="3"
        >
          3
        </option>
        <option
          value="4"
        >
          4
        </option>
      </select>
    </div>
    <div>
      <label>
        Iteration to Track
         
        <input
          className="dragon-curve-iteration-to-track"
          onChange={[Function]}
          type="text"
          value="0"
        />
      </label>
    </div>
    <div />
  </div>
</div>
`;
